DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,519 @@
+        /// Throws exceptions if properties are violated.
+        /// Useful for debugging.
+        /// </summary>
+        /// <remarks>
+        /// The properties that are checked are:
+        /// <list type="number">
+        /// <item>The level of every leaf node is one.</item>
+        /// <item>The level of every left child is exactly one less than that of its parent.</item>
+        /// <item>The level of every right child is equal to or one less than that of its parent.</item>
+        /// <item>The level of every right grandchild is strictly less than that of its grandparent.</item>
+        /// <item>Every node of level greater than one has two children.</item>
+        /// </list>
+        /// More information: https://en.wikipedia.org/wiki/AA_tree .
+        /// </remarks>
+        /// <param name="node">The node to check from.</param>
+        private void ValidateTree(AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return;
+            }
+
+            // Check level == 1 if node if a leaf node.
+            if (node.Left == null && node.Right == null)
+            {
+                if (node.Level != 1)
+                {
+                    throw new Exception($"Node {node.Key} - leaf: expected level of 1, found {node.Level} instead.");
+                }
+            }
+
+            // Check level of left child is exactly one less than parent.
+            if (node.Left != null)
+            {
+                if (node.Level - node.Left.Level != 1)
+                {
+                    throw new Exception($"Node {node.Key} - left child: expected level of {node.Level - 1}, found {node.Left.Level} instead.");
+                }
+            }
+
+            // Check level of right child is equal or one less than parent.
+            if (node.Right != null)
+            {
+                if (node.Level - node.Right.Level != 1 && node.Level != node.Right.Level)
+                {
+                    throw new Exception($"Node {node.Key} - right child: expected level of {node.Level - 1} or {node.Level}, found {node.Right.Level} instead.");
+                }
+            }
+
+            // Check right grandchild level is less than node.
+            if (node.Right != null && node.Right.Right != null)
+            {
+                if (node.Right.Level < node.Right.Right.Level)
+                {
+                    throw new Exception($"Node {node.Key} - right grandchild: right grandchild has level of {node.Right.Right.Level}, which is greater than node level of {node.Level}.");
+                }
+            }
+
+            // Check if node has two children if not leaf.
+            if (node.Level > 1)
+            {
+                if (node.Left == null || node.Right == null)
+                {
+                    throw new Exception($"Node {node.Key}: node level is greater than one but has less than two children.");
```

Comment:
Codacy found an issue: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6367315)